feat: Add folderUID and folderRef in GrafanaDashboard CRD

aboulay-numspot opened the pull request:
## Aim of the merge request

This MR aims to implement the content of the document discussed in maintainer meeting (currently under #1592).

The features are the following:
If `folder` is set and none of the other fields is set (backwards compat. case): Create the folder
If `folderUID` or `folderRef` is set, don't create the folder
If `folderUID` or `folderRef` AND folder is set, the new fields take priority -> don't create the folder

## Breaking change

There is no breaking change. We implement on the top of existing GrafanaDashboard CR behavior.

aboulay-numspot commented:
@theSuess - I know you are interested about this implementation. If it can help you to propose something even better, feel free to propose an evolution in a dedicated MR and close this one.

theSuess commented:
Waiting for https://github.com/grafana/grafana-operator/pull/1604 to be merged first as it introduces shared logic we can reuse here

aboulay-numspot commented:
@theSuess I have updated the GrafanaDashboard CR to match with the interface. To do this, I have added a Condition block into the GrafanaDashboard.Status and add the state update into the Reconciliation loop. I removed the duplicated code with the interface developed in the GrafanaFolder MR (#1604).

Can you have a look on it? I don't understand the full meaning of the Condition object, I would be happy if someone can give me some feedback into this implementation.

I think I'm ready for a review this time.

theSuess commented on api/v1beta1/grafanadashboard_types.go:
```diff
@@ -85,6 +87,14 @@ type GrafanaDashboardSpec struct {
 
+	// find parent folder for dashboard
+	// +optional
+	FolderUID string `json:"folderUID" validate:"omitempty,uuid4"`
```
Grafana Folder UIDs are not necessarily uuid4 and can be arbitrary strings. I'd get rid of this validation rule here

aboulay-numspot replied:
Handled

theSuess commented on api/v1beta1/grafanadashboard_types.go:
```diff
@@ -85,6 +87,14 @@ type GrafanaDashboardSpec struct {
 	// +optional
 	FolderTitle string `json:"folder,omitempty"`
 
+	// find parent folder for dashboard
```
```suggestion
	// UID of the target folder for this dashboard
```

aboulay-numspot replied:
Handled

theSuess commented on api/v1beta1/grafanadashboard_types.go:
```diff
@@ -85,6 +87,14 @@ type GrafanaDashboardSpec struct {
+	// +optional
+	FolderUID string `json:"folderUID" validate:"omitempty,uuid4"`
+
+	// find parent folder using GrafanaFolder CR reference
```
```suggestion
	// Name of a `GrafanaFolder` resource in the same namespace
```

aboulay-numspot replied:
Ok I removed this one

aboulay-numspot replied:
Handled

theSuess commented on controllers/dashboard_controller.go:
```diff
@@ -56,8 +57,10 @@ import (
+	initialSyncDelay               = "10s"
+	syncBatchSize                  = 100
+	conditionDashboardSynchronized = "DashboardSynchronized"
+	conditionErrorFetchingInstance = "ErrFetchingInstances"
```
Not sure if this should be a condition as the error is not fixable by the end user - only by kubernetes/operator admins

aboulay-numspot replied:
Ok I removed this one

theSuess reviewed (changes requested):
Thanks for updating this! I found a few minor nitpicks but looks great overall!
